invenio_db/uow.py

```diff
@@ -193,9 +197,17 @@ def commit(self):
             op.on_post_commit(self)
         self._mark_dirty()
 
-    def rollback(self):
+    def rollback(self, exception=None):
         """Rollback the database session."""
         self.session.rollback()
+
+        # Run exception operations
+        for op in self._operations:
+            op.on_exception(self, exception)
+
+        # Commit exception operations
+        self.session.commit()
```

Comment:
to be sure, what should that commit commit to the database and what should remain to rollback? in my understanding after that commit nothing is more present in the transaction to be rolledback

Reply:
The `on_exception` handlers are the ones that might be performing some clean-up operations that then need to be committed.

The use case we have is:

- During record publish, we want to perform content spam checks (inside a service component)
- If the checks detect spam content, we want to stop the publish from happening
- The only way to do this would be to raise an Exception inside the component method
- We also want to block the user though, and to do that we would still need to commit to the DB.
- Since we can't just keep "parts" of the UoW to the DB, we need this new hook so that there's room to run some "clean-up" code that will get committed.

Reply:
I see a couple of alternatives:

- using `on_(post_)rollback`:
  - Send off a Celery task there to perform the clean-up action
  - Register an operation that itself runs `db.session.commit()`
- do something in error handlers based on the type exception that's raised (i.e. bake in the clean-up in the raised exception or its handler)

The overall issue is that it's difficult to affect the control flow when inside a service call.

Reply:
i see potentially problems with that approach if we migrate to sqlalchemy >= 2.0. as i understand sqlalchemy 2.0 the commit stores all information inside of the transaction to the database. so if the plan is to only commit what has been added by the `on_exception` hook we have to check how to do that.

i may be seeing problems where there are none, but sqlalchemy >= 2.0 is handling the transactions differently. 

could you please write a test for that so we can test it afterwards with sqlalchemy 2.0

Reply:
Added a test (for the entire unit of work), which also does some DB operations and relies on things being rolled back properly.

The test is pretty basic, but if you have a more complex example/case in mind I can integrate it as well